server/coordinator/procedure/shard/shard_fsm.go

```diff
@@ -0,0 +1,229 @@
+// Copyright 2022 CeresDB Project Authors. Licensed under Apache-2.0.
+
+package shard
+
+import (
+	"context"
+
+	"github.com/CeresDB/ceresdbproto/pkg/clusterpb"
+	"github.com/CeresDB/ceresmeta/server/cluster"
+	"github.com/CeresDB/ceresmeta/server/coordinator/procedure/dispatch"
+	"github.com/looplab/fsm"
+	"github.com/pkg/errors"
+)
+
+// Shard FSM Const Definition
+// It contains the event name and the state name
+const (
+	StateLeader          = "LEADER"
+	StateFollower        = "FOLLOWER"
+	StatePendingLeader   = "PENDING_LEADER"
+	StatePendingFollower = "PENDING_FOLLOWER"
+
+	EventPrepareTransferFollower = "PrepareTransferFollower"
+	EventTransferFollower        = "TransferFollower"
+	EventTransferFollowerFailed  = "TransferFollowerFailed"
+
+	EventPrepareTransferLeader = "PrepareTransferLeader"
+	EventTransferLeader        = "TransferLeader"
+	EventTransferLeaderFailed  = "TransferLeaderFailed"
+)
+
+// Declare the source state array of FSM, avoid creating arrays repeatedly every time you create an FSM
+var (
+	leaderFsmEvent = fsm.Events{
+		{Name: EventPrepareTransferFollower, Src: []string{StateLeader}, Dst: StatePendingFollower},
+		{Name: EventTransferFollower, Src: []string{StatePendingFollower}, Dst: StateFollower},
+		{Name: EventTransferFollowerFailed, Src: []string{StatePendingFollower}, Dst: StateLeader},
+	}
+	leaderFsmCallbacks = fsm.Callbacks{
+		EventPrepareTransferFollower: prepareTransferFollowerCallback,
+		EventTransferFollower:        transferFollowerCallback,
+		EventTransferFollowerFailed:  transferFollowerFailedCallback,
+	}
+)
+
+var (
+	followerFsmEvent = fsm.Events{
+		{Name: EventPrepareTransferLeader, Src: []string{StateFollower}, Dst: StatePendingLeader},
+		{Name: EventTransferLeader, Src: []string{StatePendingLeader}, Dst: StateLeader},
+		{Name: EventTransferLeaderFailed, Src: []string{StatePendingLeader}, Dst: StateFollower},
+	}
+	followerFsmCallbacks = fsm.Callbacks{
+		EventPrepareTransferLeader: prepareTransferLeaderCallback,
+		EventTransferLeader:        transferLeaderCallback,
+		EventTransferLeaderFailed:  transferLeaderFailed,
+	}
+)
+
+type LeaderCallbackRequest struct {
+	Ctx            context.Context
+	C              *cluster.Cluster
+	ActionDispatch dispatch.ActionDispatch
+
+	OldLeaderNode    string
+	OldLeaderShardID uint32
```

Comment:
How about just calling it `Node` and `ShardID`? For the FSM self, there is indeed no other shard.